kairos-server/src/config.rs

```diff
@@ -1,26 +1,106 @@
-            val.parse::<T>()
-                .map_err(|e| format!("Failed to parse {}: {}", env, e))
-        })
+    let s = String::deserialize(deserializer)?;
+    match s.to_lowercase().as_str() {
+        "error" => Ok(Level::ERROR),
+        "warn" => Ok(Level::WARN),
+        "info" => Ok(Level::INFO),
+        "debug" => Ok(Level::DEBUG),
+        "trace" => Ok(Level::TRACE),
+        _ => Ok(Level::INFO),
+    } 
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Server {
+    pub address: String,
+    pub port: u16,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Log {
+    #[serde(deserialize_with = "deserialize_level_filter", serialize_with = "serialize_level_filter")]
+    pub level: Level,
+    pub log_file: Option<String>,
+}
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct Config {
+    pub server: Server,
+    pub log: Log,
 }
+
+impl Config {
+    pub fn socket_address(&self) -> SocketAddr {
+        format!("{}:{}", self.server.address, self.server.port)
+            .parse()
+            .expect("Invalid address")
+    }
+
+    pub fn initialize_logger(&self) {
+        let stdout_log = tracing_subscriber::fmt::layer().pretty();
+        let level_filter = tracing_subscriber::EnvFilter::new(&self.log.level.to_string());
+        let subscriber = tracing_subscriber::Registry::default().with(stdout_log).with(level_filter);
+
+        let file_log = if let Some(file_name) = &self.log.log_file  {
+            let file = File::create(file_name).expect("Failed to open log file, check filesystem permissions and working directory.");
+            Some(tracing_subscriber::fmt::layer()
+                .with_file(true)
+                .with_line_number(true)
+                .with_writer(file)
+            )
+        } else {
+            None
+        };
+
+        let subscriber = subscriber.with(file_log);
+            
+
+        tracing::subscriber::set_global_default(subscriber).expect("Unable to set global logging configuration");
+    }
+
+    pub fn read_config() -> Config {
+        let config_file = "kairos_config.toml";
+        if !std::path::Path::new(config_file).exists() {
+            let default_config = Config {
+                server: Server {
+                    address: "0.0.0.0".to_string(),
+                    port: 8080,
+                },
+                log: Log {
+                    level: Level::TRACE,
+                    log_file: None,
+                }
+            };
+            let toml_string = toml::to_string_pretty(&default_config).unwrap();
+            std::fs::write(config_file, toml_string).expect("Failed to write config file, file permissions might be incorrect.");
+        }
+        let config: Config = toml::from_str(
+            &std::fs::read_to_string(config_file).expect("Failed to read config file, file permissions might be incorrect.")
+        ).expect("Failed to parse config file");
+
+        config
+    }
```

Comment:
In general apps should not create state unnecessarily.
This can be an issue when running in readonly filesystems.
ENVs are used to avoid this problem, they also play better with docker and nix.

Reply:
Would env variable/dotenv overrides and not panicking if it can't write the config file be a good compromise?

Reply:
No servers should not be creating state.
I'd be fine with reading a config file, but we should never write one, unless told to by the user.
And if we are reading a config you have to ask where, and then we get into xdg which is more for apps anyway.

ENVs are the industry standard for a reason.
`dotenvy` turns a set of ENVs into a config file in the local directory (just like kairos_config.toml).